DQMServices/Core/interface/DQMStore.h

```diff
@@ -189,440 +420,278 @@ namespace dqm::dqmstoreimpl {
-      // class. We therefore need to store a pointer to the main
-      // DQMStore instance (owner_).
-      DQMStore* owner_;
-    };  // IBooker
+      //
+      // all non-template interfaces are virtual.
+      //
+
+      virtual MonitorElementData::Scope setScope(MonitorElementData::Scope newscope);
+      // RAII-Style guard to set and reset the Scope.
+      template <MonitorElementData::Scope SCOPE>
+      struct UseScope {
+        IBooker& parent;
+        MonitorElementData::Scope oldscope;
+        UseScope(IBooker& booker) : parent(booker) { oldscope = parent.setScope(SCOPE); }
+        ~UseScope() { parent.setScope(oldscope); }
+      };
+      using UseLumiScope = UseScope<MonitorElementData::Scope::LUMI>;
+      using UseRunScope = UseScope<MonitorElementData::Scope::RUN>;
+      using UseJobScope = UseScope<MonitorElementData::Scope::JOB>;
+
+      ~IBooker() override;
+
+    protected:
+      IBooker(DQMStore* store);
+      virtual uint64_t setModuleID(uint64_t moduleID);
+      virtual edm::LuminosityBlockID setRunLumi(edm::LuminosityBlockID runlumi);
+      virtual MonitorElement* bookME(TString const& name,
+                                     MonitorElementData::Kind kind,
+                                     std::function<TH1*()> makeobject,
+                                     bool forceReplace = false);
+
+      DQMStore* store_ = nullptr;
+      MonitorElementData::Scope scope_ = MonitorElementData::Scope::JOB;
+      uint64_t moduleID_ = 0;
+      edm::LuminosityBlockID runlumi_ = edm::LuminosityBlockID();
+    };
 
-    class IGetter {
+    class IGetter : public dqm::implementation::NavigatorBase {
     public:
-      friend class DQMStore;
+      // The IGetter interface is not really suitable for concurrent lumis/runs,
+      // so we don't even try much to get it right. Harvesting needs to run
+      // sequentially for now. Therefore, the methods just return the next-best
+      // global MEs that they find, ignoring Scope and run/lumi.
+      // Since these are global MEs, they may be deleted at some point; don't
+      // store the pointers!
+      // They are also shared with other modules. That is save when running
+      // multi-threaded as long as getTH1() etc. are not used, but of course
+      // all dependencies need to be properly declared to get correct results.
+
+      // TODO: review and possibly rename the all methods below:
+      // get MEs that are direct children of full path `path`
+      virtual std::vector<dqm::harvesting::MonitorElement*> getContents(std::string const& path) const;
+
+      // get all elements below full path `path`
+      // we have to discuss semantics here -- are run/lumi ever used?
+      virtual std::vector<dqm::harvesting::MonitorElement*> getAllContents(std::string const& path) const;
+      DQM_DEPRECATED
+      virtual std::vector<dqm::harvesting::MonitorElement*> getAllContents(std::string const& path,
```

Comment:
No, in general harvesting needs to modify objects at any time. This is one of the reasons why we don't save MEs in products yet: Immutable MEs require nontrivial changes to lots of harvesting logic.

Reply:
Would the opposite, i.e. not declaring these functions `const`, work either?

Reply:
Potentially, though I am not sure I support your concern: `getAllContents` is const on the `DQMStore` (which is important: the datastructure in `DQMStore` does not allow concurrent access etc.), while what you do the `ME`s is completely up to you (e.g. they support concurrent filling). There are some evil interconnections, e.g. around the `Key` of the `ME`, which are mostly because this needs to preserve most of the old API.

I'll give this a closer look tomorrow.

Reply:
I'm merely thinking about our code rule 4.23 "a `const` member function must not give non-const access to any data directly or indirectly held by the object". (ok, this case might be one of those where one has to compromise, at least in the short term)

> Potentially, though I am not sure I support your concern: `getAllContents` is const on the `DQMStore` (which is important: the datastructure in `DQMStore` does not allow concurrent access etc.)

But I don't quite follow why "datastructure in `DQMStore`" not allowing concurrent access implies that `getAllContents()` is good to be `const`?

Reply:
> But I don't quite follow why "datastructure in DQMStore" not allowing concurrent access implies that getAllContents() is good to be const?

It doesn't really, but there is many places where const-ness got very weird in this design, due to the constraint that in event processing, the `DQMStore` must not be mutated while the MEs will be, while in harvesting, both can be modified. I tried hard to get everything to work without const-casts (and almost succeeded), I didn't care too much wat makes sense.

I'll have a closer look if this const-ness can be changed, that depends on how much code is affected by such an API change.

Reply:
Ok. Given that the situation wrt. "leaking" non-`const` pointers from `const` functions probably was there already, I'm fine with leaving further investigation for later time (also given the size of this PR and further work depending on it). So let's say this is something to think about.